ctapipe/tools/apply_models.py

```diff
@@ -100,84 +138,72 @@ def setup(self):
-            self.h5file.close()
-            self.h5file = self.enter_context(
-                tables.open_file(self.output_path, mode="r+")
-            )
-            self.loader.h5file = self.h5file
-
-    def _apply(self, reconstructor):
-        prefix = reconstructor.prefix
-
-        desc = f"Applying {reconstructor.__class__.__name__}"
-        unit = "chunk"
-
         chunk_iterator = self.loader.read_telescope_events_by_id_chunked(
             self.chunk_size
         )
-        for start, stop, chunk in tqdm(chunk_iterator, desc=desc, unit=unit):
-            tel_tables = []
-
-            for tel_id, table in chunk.items():
-                tel = self.loader.subarray.tel[tel_id]
-                if tel not in reconstructor._models:
-                    self.log.warning(
-                        "No model in %s for telescope type %s, skipping tel %d",
-                        reconstructor,
-                        tel,
-                        tel_id,
-                    )
-                    continue
-
-                if len(table) == 0:
-                    self.log.warning("No events for telescope %d", tel_id)
-                    continue
-
-                table.remove_columns(
-                    [c for c in table.colnames if c.startswith(prefix)]
-                )
+        bar = tqdm(
+            chunk_iterator,
+            desc="Applying reconstructors",
+            unit=" Array Events",
+            total=chunk_iterator.n_total,
+            disable=not self.progress_bar,
+        )
+        with bar:
+            for chunk, (start, stop, tel_tables) in enumerate(chunk_iterator):
+                for reconstructor in self._reconstructors:
+                    self.log.debug("Applying %s to chunk %d", reconstructor, chunk)
+                    self._apply(reconstructor, tel_tables, start=start, stop=stop)
 
-                predictions = reconstructor.predict_table(tel, table)
+                bar.update(stop - start)
 
-                for prop, prediction_table in predictions.items():
-                    new_columns = prediction_table.colnames
-                    output_columns = ["obs_id", "event_id", "tel_id"] + new_columns
+    def _apply(self, reconstructor, tel_tables, start, stop):
+        prefix = reconstructor.prefix
 
-                    # copy columns into full feature table
-                    for col in new_columns:
-                        table[col] = prediction_table[col]
+        for tel_id, table in tel_tables.items():
+            tel = self.loader.subarray.tel[tel_id]
 
-                    write_table(
-                        table[output_columns],
-                        self.output_path,
-                        f"/dl2/event/telescope/{prop}/{prefix}/tel_{tel_id:03d}",
-                        append=True,
-                    )
-                tel_tables.append(table)
+            if tel not in reconstructor._models:
+                self.log.warning(
```

Comment:
Should this be a warning or raise an error?
I am not entirely sure yet what I would expect to happen if my reconstructor has not been trained for the complete array.

Reply:
It should probably be an error, yes. Will have to check how that works with our test files though.

Reply:
Since this is not a change to before, let's discuss in an issue